### PR TITLE
Add additional tests, functionality for optional attendees

### PR DESCRIPTION
**Pull request by ganewman**
Oh god, it did the thing where it included the last commit as well. I don't know how to fix it. Only the last two should be included.

Add additional tests as specified by the walkthrough, as well as the added functionality to pass these tests. Optional attendees will now be considered when attempting to schedule events.

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,120 @@
...
+     * 1. Get a list of all of the events which actually concern the attendees of the request.
+     * 2. Sort the list by start time.
+     * 3. Loop through, and find all available times "between" these events; requires some amount
+     * of shifting around bounds to deal with edge cases regarding nested events.
+     * 4. Return a list of these times, only adding a time if it is longer than the proposed
+     * duration of the MeetingRequest. 
+     */
+     
+    // Run the algorithm the first time also including optional attendees
+  	Collection<Event> relevantEvents = getAllRelevantEvents(events, request, true);
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:36]: fix the indentation. Also what is true? If you don't want to create a temp variable for include optional change this to 
getAllRelevantEvents(events, request, /*includeOptional=*/ true);
so the reader can know what true indicates

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,120 @@
...
+    return proposedSolution;
+  }
+
+  /**
+   * Returns a Collection of only those events which have
+   * attendees specified in the request parameter.
+   */
+  private Collection<Event> getAllRelevantEvents(Collection<Event> events, MeetingRequest request,
+  	  boolean includeOptional) {
+    // Make a copy of the attendees list which can be modified to add optional attendees if need be
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: nit. Add a period.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,120 @@
...
+    }
+    return relevantEvents;
+  }
+
+  /**
+  * Returns a sorted (by start time) List<TimeRange> representing the Collection of events
+  * passed as a parameter. 
+  */
+  private List<TimeRange> getSortedTimeRanges(Collection<Event> events) {
+    List<TimeRange> times = new ArrayList<>();
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:78]: Initialize this to have the same capacity as events.size()

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,120 @@
...
+    for (Event event : events) {
+      times.add(event.getWhen());
+    }
+    Collections.sort(times, TimeRange.ORDER_BY_START);
+    return times;
+  }
+
+  /** Returns a Collection of only those times which do not overlap with any TimeRange in times */
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:86]: nit. Add period.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,120 @@
...
+	private Collection<TimeRange> getComplementTimes(List<TimeRange> times, MeetingRequest request) {
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:87]: Fix indentation for this function

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,120 @@
...
+    Collection<TimeRange> complementTimes = new ArrayList<>();
+    if (times.size() == 0) {
+      if (proposedTimeIsLongEnough(TimeRange.WHOLE_DAY, request)) {
+        complementTimes.add(TimeRange.WHOLE_DAY);
+      }
+      return complementTimes;
+	  }
+    // Add any time before the first event in the list
+    if (! (TimeRange.START_OF_DAY == times.get(0).start())) {
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:96]: You can switch this to !=

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 7]
```diff
@@ -121,6 +123,65 @@ public void everyAttendeeIsConsidered() {
     Assert.assertEquals(expected, actual);
   }
 
+    @Test
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:126]: Fix indentation for your test annotations. They should line up with the functions

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -31,17 +31,35 @@
...
    */
-  private Collection<Event> getAllRelevantEvents(Collection<Event> events, MeetingRequest request) {
-    Collection<String> attendees = request.getAttendees();
+  private Collection<Event> getAllRelevantEvents(Collection<Event> events, MeetingRequest request,
+  	  boolean includeOptional) {
+    // Make a copy of the attendees list which can be modified to add optional attendees if need be
+    Collection<String> attendees = new HashSet<>(request.getAttendees());
+    if (includeOptional) {
+      Collection<String> optionalAttendees = request.getOptionalAttendees();
+      attendees.addAll(optionalAttendees);
```
Comment (kellyktan) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:61]: nit: Can probably just do this in one line: `attendees.addAll(request.getOptionalAttendees());`

File: walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java [hunk 9]
```diff
@@ -121,6 +123,65 @@ public void everyAttendeeIsConsidered() {
...
+
+    Assert.assertEquals(expected, actual);
+  }
+
+    @Test
+  public void everyAttendeeIsConsideredPlusOptional() {
+    // Have each person have different events. We should see two options because each person has
+    // split the restricted times.
+    //
+    // Events  :       |--A--|--C--|--B--|
```
Comment (suparnabarua) [inline walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:161]: nit. in your diagrams, can you make the optional attendees a different line (similar to the above test)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (ganewman) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Can you do that in Java?

Comment (ganewman) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'll just make a temporary variable.

**Pull request merged**